xmppserver/src/main/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java

```diff
@@ -1621,7 +1621,36 @@ public void markedAsSeniorClusterMember() {
     private void restoreCacheContent()
     {
         Log.debug( "Restoring cache content for cache '{}' by adding all outgoing server routes that are connected to the local cluster node.", serversCache.getName() );
-        localRoutingTable.getServerRoutes().forEach( route -> route.getOutgoingDomainPairs().forEach( address -> serversCache.put( address, server.getNodeID()) ) );
+
+        // Check if there are local s2s connections that are already in the cache for remote nodes
+        Set<DomainPair> localServerRoutesToRemove = new HashSet<>();
+        localRoutingTable.getServerRoutes().forEach(
+            route -> route.getOutgoingDomainPairs().forEach(
+                address -> {
+                    final Lock lock = serversCache.getLock(address);
+                    lock.lock();
+                    try {
+                        if (serversCache.containsKey(address)) {
+                            Log.info("We have an s2s connection to {}, but this connection also exists on other nodes. They are not allowed to both exist, so this local s2s connection will be terminated.", address);
+                            localServerRoutesToRemove.add(address);
+                        } else {
+                            serversCache.put(address, server.getNodeID());
+                        }
+                    } finally {
+                        lock.unlock();
+                    }
+                })
+        );
+        for (DomainPair localServerRouteToRemove : localServerRoutesToRemove) {
+            final RoutableChannelHandler route = localRoutingTable.getRoute(localServerRouteToRemove);
```

Comment:
Is it worth wrapping this in a `try/catch` to prevent closure of any one route to stop all remaining routes from being closed?

Reply:
Yes.